src/bridge/cockpitcpusamples.c

```diff
@@ -132,48 +132,65 @@ read_file (const gchar *path)
 }
 
 static void
-sample_cpu_sensors (CockpitSamples *samples,
-                    int hwmonID)
+sample_cpu_sensors (gchar *sensor_path,
+                    CockpitSamples *samples)
+{
+    g_autofree gchar *temp_content = read_file (sensor_path);
+
+    if (temp_content == NULL)
+        return;
+
+    gint64 temperature = g_ascii_strtoll (temp_content, NULL, 10);
+    // overflow or invalid
+    if (temperature == G_MAXINT64 || temperature == G_MININT64 || temperature == 0)
+      {
+        g_debug("Invalid number in %s: %m", sensor_path);
+        return;
+      }
+
+    cockpit_samples_sample (samples, "cpu.temperature", sensor_path, temperature/1000);
+}
+
+static void
+add_cpu_sensor_files (GList **devices,
+                      gchar *name,
+                      int hwmonID)
 {
   for (int i = 1; TRUE; i++)
     {
-      g_autofree gchar *temp = g_strdup_printf ("/sys/class/hwmon/hwmon%d/temp%d_input", hwmonID, i);
-      g_autofree gchar *temp_content = read_file (temp);
-
-      // end loop on error
-      if (temp_content == NULL)
+      // break loop if input file does not exist
+      g_autofree gchar *sensor_path = g_strdup_printf ("/sys/class/hwmon/hwmon%d/temp%d_input", hwmonID, i);
+      if (!g_file_test (sensor_path, G_FILE_TEST_EXISTS))
           break;
 
-      gint64 temperature = g_ascii_strtoll (temp_content, NULL, 10);
-      // overflow or invalid
-      if (temperature == G_MAXINT64 || temperature == G_MININT64 || temperature == 0)
-        {
-          g_debug("Invalid number in %s: %m", temp);
-          continue;
-        }
-
       g_autofree gchar *label = g_strdup_printf ("/sys/class/hwmon/hwmon%d/temp%d_label", hwmonID, i);
       g_autofree gchar *label_content = read_file (label);
 
       if (label_content == NULL)
         {
           // labels aren't used on ARM
-          label_content = g_strdup_printf ("Core %d", i);
+          if (!g_str_equal (name, "cpu_thermal"))
+              continue;
+        }
+      else
+        {
+          g_strchomp (label_content);
+
+          // only sample CPU Temperature in atk0110
+          if (!g_str_equal (label_content, "CPU Temperature") && g_str_equal (name, "atk0110"))
+              continue;
+          // ignore Tctl on AMD devices
+          if (g_str_equal (label_content, "Tctl"))
+              continue;
         }
 
-      g_strchomp (label_content);
-
-      // ignore Tctl on AMD devices
-      if (g_str_equal (label_content, "Tctl"))
-          continue;
-
-      g_autofree gchar *instance = g_strdup_printf ("hwmon%d %s", hwmonID, label_content);
-      cockpit_samples_sample (samples, "cpu.temperature", instance, temperature/1000);
+      if (devices != NULL)
```

Comment:
This shouldn't be a runtime check -- if it's NULL, it's a programming error. Please make this an assertion instead at the top of the function.

You can also simplify this by just creating a devices list from scratch and returning it, instead of getting passed a pointer to the list pointer. (That's a matter of taste, I'm fine with either.)